### PR TITLE
Add support for adding custom query parameters to URL

### PR DESCRIPTION
**Pull request by remyleone**
##### SUMMARY

It is currently not possible to specify query parameters to Scaleway URL. For instance, it is not currently possible to specify how many results you want per page. It makes it difficult to test pagination for Scaleway compute without hitting the default number of machines per pages (50).

By adding arbitrary queries parameters to each ansible modules we will be able to have a greater flexibility in our API calls.

Will be very useful to test #46117

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

- scaleway

##### ANSIBLE VERSION

```
ansible 2.8.0.dev0 (scaleway_query_filters 2c9caa04d2) last updated 2018/10/02 12:53:33 (GMT +200)
  config file = None
  configured module search path = [u'/Users/sieben/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/sieben/workspace/ansible/lib/ansible
  executable location = /Users/sieben/workspace/ansible/bin/ansible
  python version = 2.7.15 (default, Jun 17 2018, 12:46:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```

### CODE REVIEW

**Comment by remyleone** [comment]
@Spredzy @pilou- could I get a review?

**Comment by remyleone** [comment]
@gundalow Could you review and merge?

**Comment by remyleone** [comment]
@pilou- Could you review?

**Comment by mscherer** [comment]
So, I am a bit concerned about the level of code duplication between module_utils/scaleway.py and the online.py one, especially since we start to change one to add more flexibility that could end being useful to the others. And also concerned about cut and paste with DO module_utils.

**Comment by mscherer** [comment]
shipit

**Comment by remyleone** [comment]
shipit

**Comment by pilou-** [comment]
I have a negative opinion on this PR:
- pagination settings must not be module parameters instead use something like `max_results` 
- if a user really would to do that, he could use `uri` module

**Comment by remyleone** [comment]
@pilou- This PR is not about pagination. This PR is about adding the possibility to add query_parameters to the URL from the module. This can be used to offer the possibility to the user to add optional filters to speed up queries for instance if they want to and if the module does not support it natively.

**Comment by pilou-** [comment]
Users still can use URI module :) I am not convinced by the given examples. And what about the idempotence ? Should not this feature be used only for facts modules ?

**Comment by remyleone** [comment]
I don't understand what idempotence have to do with this PR. I also don't understand why it should also only be used for facts? Could you give examples?

File: lib/ansible/utils/module_docs_fragments/scaleway.py [hunk 1]
```diff
@@ -22,6 +22,11 @@ class ModuleDocFragment(object):
       - HTTP timeout to Scaleway API in seconds.
     default: 30
     aliases: ['timeout']
+  query_parameters:
+    description:
+    - List of parameters passed to the query string
```
Comment (halberom) [inline lib/ansible/utils/module_docs_fragments/scaleway.py:27]: should be a dict of parameters and `type: dict`

**Comment by remyleone** [comment]
@halberom How do you handle the case where you have several time the same key in a url query string? Espically with `urlencode`?

**Comment by halberom** [comment]
@sieben, not sure what you mean. I made my comment because it looked like you had typos on the doc of the param. It states `type: list` but has `default: {}`. If it should be a list (of dicts), then your default is wrong (and possibly your examples and code).

File: lib/ansible/utils/module_docs_fragments/scaleway.py (hunk 1 above)
Comment (remyleone) [reply lib/ansible/utils/module_docs_fragments/scaleway.py:27]: fixed

**Comment by remyleone** [comment]
@gundalow Could this be merged?

**Comment by gundalow** [comment]
Merged into `devel` for release in Ansible 2.8.

Thanks for the PR and reviews!

**Pull request merged**